### ISSUE TITLE
Improve utility-function loading

### ISSUE THREAD

**Issue by dritter** [bhilburn/powerlevel9k#146]
Disclaimer: This is by now untested. For now this issue is just a reminder for me (or anybody who is interested in testing this).

Some users load the theme with [zgen](https://github.com/tarjoilija/zgen) (like [tritlo](https://github.com/Tritlo/dotfiles/blob/196760a0fe7a68a8901c448e0757f57cac6df5ba/zsh/.zshrc#L47-L48)). This might be a problem, if zgen loads the theme as a function (like prezto does). 

So we need a solution to find the file in which a function was defined, to get the base path of our theme. Or we make this configurable by a variable, so users could work around that problem.

**Comment by Tritlo** [comment]
Wow, thanks. My solution is a bit of a hack, but I want my .zshrc to be as self contained as possible.

**Comment by dritter** [comment]
@Tritlo I think you got me wrong here. This is not about you, writing your dotfiles wrong. It is about how _powerlevel9k_ loads utility functions. That does not play well, if the theme is loaded as a function. In that case we assume that the theme was loaded by prezto and seek for the utility files in a prezto-like path. [This](https://github.com/bhilburn/powerlevel9k/blob/f9957ca5592df0ab992d3750e18ee6f61b3e3002/powerlevel9k.zsh-theme#L20-L25) is the - probably - faulty line.

Another example is [outcoldman](https://outcoldman.com/en/archive/2015/09/13/keep-your-sh-together/). He loads the theme also by zgen.

Did not see that coming.. :cry:

**Comment by bhilburn** [comment]
@dritter - Agreed. The issue here is the design we currently have in `next` where we have broken out utility functions. We need to figure out how to make this work to make workflows like @Tritlo easier.

@Tritlo - Sorry for any miscommunication, here! Would you be up for helping us test a fix to powerlevel9k to make this work better?

**Comment by Tritlo** [comment]
@dritter I see. Should I rather just use

    zgen load bhilburn/powerlevel9k

how does this affect the functionality of powerlevel9k? Is it only the next branch?

**Comment by dritter** [comment]
@Tritlo No no. All of the different methods should work. Regardless how you load the theme and regardless if the theme was symlinked or loaded directly.

And yes, it is currently only on `next` branch. So the question could be answered easily: Does the `next` branch work for you? Or do you get weird error messages?

**Comment by Tritlo** [comment]
Yes, the branch seems to work, and no weird errors. However, the battery segment does not seem to work.

**Comment by rjorgenson** [comment]
@Tritlo What issue are you running into with the battery segment? I've had limited environments to test it on(only ubuntu 14.04 and OS X) and any insight you could provide would be very helpful to making it more portable =]

**Comment by dritter** [comment]
@bhilburn I think we should have this in mind even `zgen` seems to be clear. The hardcoded path bothers me..

**Comment by Tritlo** [comment]
@rjorgenson Apparently the battery plugin works for me (ubuntu 14.04), I just had the color setting so that the text was black on black, making it invisible :/

**Comment by rjorgenson** [comment]
Ahh, well that's good that it works =] I'll take a look at it and see if I can figure out a better way to set default colors so that doesn't happen.

**Comment by Tritlo** [comment]
@rjorgenson you could maybe just swap out "black" for "cyan". I doubt many people have cyan as their default color.

**Comment by dritter** [comment]
To stay on topic: `command -V` could be our friend:

```
$ command -V get_icon_names
get_icon_names is a shell function from /Users/dr/.oh-my-zsh/custom/themes/powerlevel9k-wip/functions/icons.zsh
```
That works for me in ZSH 5.1.1.

**Comment by dritter** [comment]
Some news here:
`command -V` is just an alias to `whence -v`. And `whence -v` prints the path to the script since ZSH 5.0.8 ([link](http://zsh.sourceforge.net/releases.html)). The origin of that feature seems to lie in [this message](http://www.zsh.org/mla/workers/2015/msg01096.html).
I haven't found another way of getting the installation path automatically. In the case that this is still not sufficient, I added a variable `POWERLEVEL9K_INSTALLATION_PATH`, which needs to be set to the exact location (including the file name).

I've made pull request #154 to address this.
